### PR TITLE
[REVIEW][FEATURE][UWANG-170] - Integration signin admin

### PR DESCRIPTION
**Pull request by fahmi-azzuhri**
[REVIEW][FEATURE][UWANG-170] - Integration signin admin

#### Description

> Description of features, bugs, improvements or any changes in this MR.

#### Ticket Reference(s)

[UWANG-170](https://bluehabit.atlassian.net/browse/UWANG-170)

##### UI Changes

> Screenshot or video of UI-related changes

##### Coverage

> Test coverage screenshot.

N/A

#### Note(s)

> Mention other user or extra note unrelated to the MR here.

[UWANG-170]: https://bluehabit.atlassian.net/browse/UWANG-170?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[UWANG-170]: https://bluehabit.atlassian.net/browse/UWANG-170?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: components/SigninForm.vue [hunk 1]
```diff
@@ -1,7 +1,13 @@
 <script setup lang="ts">
 import { ref } from 'vue'
+import { useAuth } from '../store/useAuth'
```
Comment (triandamai) [inline components/SigninForm.vue:3]: gausah pake import ya

File: components/SigninForm.vue [hunk 2]
```diff
@@ -15,6 +21,9 @@ const closeModal = () => {
     showModal.value = false
 }
 
+const handleSubmit = async () => {
+    await signIn(email.value, password.value)
```
Comment (triandamai) [inline components/SigninForm.vue]: pake function aja ya

File: store/useAuth.ts [hunk 3]
```diff
@@ -1,10 +1,46 @@
-export const useAuth = defineStore("auth", {
-    state:()=>({
+import { defineStore } from 'pinia';
 
-    }),
-    actions:{
-        async singIn(){
-            //todo
+export const useAuth = defineStore('auth', {
```
Comment (triandamai) [inline store/useAuth.ts:3]: user nya di bikin type terpisah aja @fahmi-azzuhri

File: store/useAuth.ts [hunk 4]
```diff
@@ -1,10 +1,46 @@
...
+      updated_at: '',
+      roles: [],
+      profile: []
+    },
+    isLoggedIn: false,
+    errorMessage: ''
+  }),
+  actions: {
+    async signIn(email : string, password : string) {
+      try {
```
Comment (triandamai) [inline store/useAuth.ts]: gausah pake try catch ya

File: store/useAuth.ts [hunk 5]
```diff
@@ -1,10 +1,46 @@
...
+        const response = await fetch('https://api.bluhabit.id/auth/v1/admin/sign-in', {
```
Comment (triandamai) [inline store/useAuth.ts]: pake useApi ya

**Review by triandamai — CHANGES_REQUESTED** [review]
cek

**Comment by fahmi-azzuhri** [comment]
done bang, tolong di cek @triandamai

File: components/SigninForm.vue [hunk 6]
```diff
@@ -15,6 +18,9 @@ const closeModal = () => {
     showModal.value = false
 }
 
+const handleSubmit = () => {
```
Comment (triandamai) [inline components/SigninForm.vue]: pake keyword `function` ya `const,let,var` kita khususin buat variabel aja @fahmi-azzuhri

File: store/useAuth.ts [hunk 7]
```diff
@@ -1,12 +1,41 @@
...
-        async singIn(){
-            //todo
+import { defineStore } from 'pinia';
+
+const { post } = useApi();
+
+export const useAuth = defineStore('auth', {
+  state: () => ({
+    token: '',
+    user: {} as UserResponse,
```
Comment (triandamai) [inline store/useAuth.ts]: type nya jangan pake as ya tapi `defineStore<string,{},{},{}>(...)`

File: store/useAuth.ts [hunk 8]
```diff
@@ -1,12 +1,41 @@
...
+    isLoggedIn: false,
+    errorMessage: ''
+  }),
+  actions: {
+    signIn(email: string, password: string) {
```
Comment (triandamai) [inline store/useAuth.ts]: pake keyword `async ya`

File: store/useAuth.ts [hunk 9]
```diff
@@ -1,12 +1,41 @@
...
+      const signInRequest = post<{ token: string; user: any }>('auth/v1/admin/sign-in', {
```
Comment (triandamai) [inline store/useAuth.ts]: pakein type ya

File: store/useAuth.ts [hunk 10]
```diff
@@ -1,12 +1,41 @@
...
+        email,
+        password,
+      });
+
+      signInRequest.then((response) => {
```
Comment (triandamai) [inline store/useAuth.ts]: langsung aja `const response = await post<UserResponse>('/...',{})`

File: store/useAuth.ts [hunk 11]
```diff
@@ -1,12 +1,41 @@
...
+        if (response.isSuccessful) {
+          if (response.data) {
+            this.token = response.data.token;
+            this.user = response.data.user;
+            this.isLoggedIn = true;
+            console.log('Sign in successful');
```
Comment (triandamai) [inline store/useAuth.ts]: jangan lupa console.log di hapus ya

File: store/useAuth.ts [hunk 12]
```diff
@@ -1,12 +1,41 @@
...
+  }),
+  actions: {
+    signIn(email: string, password: string) {
+      const signInRequest = post<{ token: string; user: any }>('auth/v1/admin/sign-in', {
+        email,
+        password,
+      });
+
+      signInRequest.then((response) => {
+        if (response.isSuccessful) {
```
Comment (triandamai) [inline store/useAuth.ts]: coba pake metode `guard clause` 
misal

```typescript
if(!response.isSuccesfull) {
  //todo

  return /*memastikan kode berenti disini*/
}

//kode ini di eksekusi sudah dijamin isSuccesfull == true
if(!response.data){
 //todo  
 return
}
//kode ini di eksekusi sudah dijamin if(data)

...
```

File: store/useAuth.ts [hunk 13]
```diff
@@ -1,12 +1,41 @@
...
+            this.errorMessage = 'No data received from the server';
+            console.error('Sign in failed: No data received');
+          }
+        } else {
+          this.errorMessage = response.message || 'Sign in failed';
+          console.error('Sign in failed');
         }
+      });
+
+      signInRequest.catch((error) => {
```
Comment (triandamai) [inline store/useAuth.ts]: ga perlu catch ya udah di handle di `useApi`

**Review by triandamai — CHANGES_REQUESTED** [review]
repisi @fahmi-azzuhri

**Comment by fahmi-azzuhri** [comment]
di SigninForm.vue

const {signIn} = useAuth()

kenapa error mulu ya bang @triandamai

**Comment by triandamai** [comment]
boleh di attach errornya ga @fahmi-azzuhri ?

**Review by triandamai — APPROVED** [review]

**Comment by fahmi-azzuhri** [comment]
Property 'signIn' does not exist on type '_StoreWithState<string, { token: string; user: UserResponse; isLoggedIn: boolean; errorMessage: string; }, {}, {}> & { token: string; user: any; isLoggedIn: boolean; errorMessage: string; } & _StoreWithGetters<...> & PiniaCustomProperties<...> & PiniaCustomStateProperties<...>'
Screenshot 2024-03-17 195907

**Comment by triandamai** [comment]
Conflict @fahmi-azzuhri 
itu error karena di defineStore nya ga di define type action nya

`defineStore<string,{/*ini type state*/},{/* ini type getter */},{/* ini type action */}`

**Comment by triandamai** [comment]
integrasi sekalian bikin page nya ya @fahmi-azzuhri

**Review by azkifairuz — APPROVED** [review]
nice

**Pull request merged**